pkg/actors/actors.go

```diff
@@ -530,7 +533,7 @@ func (a *actorsRuntime) evaluateReminders() {
 
 	var wg sync.WaitGroup
 	for _, t := range a.config.HostedActorTypes {
-		vals, err := a.getRemindersForActorType(t)
+		vals, _, err := a.getRemindersForActorType(t)
```

Comment:
Curious about the repro case.... just making sure I understand the interaction... as you said here:

> Two instances of daprd will change the reminder's key in parallel

Is the issue that we store all of the reminders for an actor type as one set?  This means that we allow the reminders within that set to vary by reminder name or actor ID. 

And so the following is possible:

Actor of type `A` with Id `1` wants to add a reminder `coolreminder`
Actor of type `A` with Id `2` wants to add a reminder `coolreminder`

Request for `A:1` reads the collection of reminders
Request for `A:2` reads the collection of reminders
Request for `A:1` adds `id: 1, name: coolreminder` and saves state
Request for `A:2` adds `id: 2, name: coolreminder` and saves state - **overwriting the state from the previous operation** and cause the runtime to forget about `id: 1, name: coolreminder`.

Is this right? it's the classic data race that appears with TOCTOU

--- 

I'm trying to make sure I'm convinced that OCC is the right fix. 

I think the following invariants make it a good fix:

- An actor instance is single-threaded
- An actor instance cannot/does-not modify the set of reminders for another actor
- An actor instance cannot read its own list of active reminders (TOCTOU is impossible within a single actor)

Reply:
Yes, Ryan. The race condition you described is exactly what I described in the issue.

We cannot have reminder per actorid because reminders outlive the actor instances. It means that the reminder for an ActorId will determine that it should be activated. Plus, we don't have a "list" feature in the state store to retrieve all keys with a given prefix.